pep-0544.txt

```diff
@@ -0,0 +1,968 @@
+that a method or data attribute does not need to be present in a class
+implementing a protocol, but if it is present, it must conform to a specific
+signature or type. One could use a ``hasattr()`` check to determine whether
+they can use the attribute on a particular instance.
+
+Languages such as TypeScript have similar features and
+apparently they are pretty commonly used. The current realistic potential
+use cases for protocols in Python don't require these. In the interest
+of simplicity, we propose to not support optional methods or attributes.
+We can always revisit this later if there is an actual need.
+
+
+Make protocols interoperable with other approaches
+--------------------------------------------------
+
+The protocols as described here are basically a minimal extension to
+the existing concept of ABCs. We argue that this is the way they should
+be understood, instead of as something that replaces Zope interfaces,
+for example. Attempting such interoperabilities will significantly
+complicate both the concept and the implementation.
+
+
+Use assignments to check explicitly that a class implements a protocol
+----------------------------------------------------------------------
+
+In Go language the explicit checks for implementation are performed
+via dummy assignments [golang]_. Such a way is also possible with the
+current proposal. Example::
+
+  class A:
+      def __len__(self) -> float:
+          return ...
+
+  _: Sized = A()  # Error: A.__len__ doesn't conform to 'Sized'
+                  # (Incompatible return type 'float')
+
+This approach moves the check away from
+the class definition and it almost requires a comment as otherwise
+the code probably would not make any sense to an average reader
+-- it looks like dead code. Besides, in the simplest form it requires one
+to construct an instance of ``A``, which could be problematic if this requires
+accessing or allocating some resources such as files or sockets.
+We could work around the latter by using a cast, for example, but then
+the code would be ugly. Therefore we discourage the use of this pattern.
+
+
+Support ``isinstance()`` checks by default
+------------------------------------------
+
+The problem with this is instance checks could be unreliable, except for
+situations where there is a common signature convention such as ``Iterable``.
+Another potentially problematic case is assignment of attributes
+*after* instantiation::
+
+  class P(Protocol):
+      x: int
+
+  class C:
+      def initialize(self) -> None:
+          self.x = 0
+
+  c = C()
+  isinstance(c1, P)  # False
+  c.initialize()
+  isinstance(c, P)  # True
+
+  def f(x: Union[P, int]) -> None:
+      if isinstance(x, P):
+          # type of x is P here
+          ...
+      else:
+          # type of x is int here?
+          print(x + 1)
+
+  f(C())   # oops
+
+We argue that requiring an explicit class decorator would be better, since
```

Comment:
As discussed in the examples I gave, it would also make protocols more fragile and potentially surprising. My argument is that explicit here is better than implicit, even if it adds a bit of work for the programmer. Also, currently my working hypothesis is that protocols won't be very common in typical statically typed code -- they will be used every now and then, but not very frequently. If correct, the extra effort needed for adding the decorator would very little and basically insignificant. If this proves to be an incorrect hypothesis, it's much easier to add the missing implicit runtime check in the future than to remove an existing implicit runtime check, since the latter would break existing code. Basically, my reasoning is that if in doubt, leave it out from the first version until we have evidence that it would be useful.